gcc/rust/util/rust-unicode.cc

```diff
@@ -0,0 +1,328 @@
+
+typedef uint32_t codepoint_t;
+typedef std::vector<codepoint_t> string_t;
+
+template <std::size_t SIZE>
+int64_t
+binary_search_ranges (
+  const std::array<std::pair<uint32_t, uint32_t>, SIZE> &ranges,
+  uint32_t target_cp)
+{
+  if (SIZE == 0)
+    return -1;
+
+  uint32_t low, high, mid;
+  uint32_t start, end;
+  low = 0;
+  high = SIZE - 1;
+  mid = (low + high) / 2;
+  while (high - low > 1)
+    {
+      start = ranges[mid].first;
+      end = ranges[mid].second;
+      if (start <= target_cp && target_cp < end)
+	{
+	  return mid;
+	}
+      else if (end <= target_cp)
+	low = mid + 1;
+      else
+	high = mid - 1;
+      mid = (low + high) / 2;
+    }
+
+  if (ranges[mid].first <= target_cp && target_cp < ranges[mid].second)
+    return mid;
+  else
+    return -1;
+}
+
+template <std::size_t SIZE>
+int64_t
+binary_search_sorted_array (const std::array<std::uint32_t, SIZE> &array,
+			    uint32_t target)
+{
+  if (SIZE == 0)
+    return -1;
+
+  uint32_t low, high, mid;
+  low = 0;
+  high = SIZE;
+  mid = (low + high) / 2;
+  while (high - low > 1)
+    {
+      if (array[mid] <= target)
+	low = mid;
+      else
+	high = mid;
+      mid = (low + high) / 2;
+    }
+
+  if (array[mid] == target)
+    return mid;
+  else
+    return -1;
+}
+
+int
+lookup_cc (codepoint_t c)
+{
+  auto it = Rust::CCC_TABLE.find (c);
+  if (it != Rust::CCC_TABLE.end ())
+    return it->second;
+  else
+    // Starter. Returns zero.
+    return 0;
+}
+
+tl::optional<codepoint_t>
+lookup_recomp (codepoint_t starter, codepoint_t c)
+{
+  auto it = Rust::RECOMPOSITION_MAP.find ({starter, c});
+  if (it != Rust::RECOMPOSITION_MAP.end ())
+    return {it->second};
+
+  it = Rust::RECOMPOSITION_MAP.find ({starter, 0});
+  if (it != Rust::RECOMPOSITION_MAP.end ())
+    return {it->second};
+
+  return tl::nullopt;
+}
+
+void
+recursive_decomp_cano (codepoint_t c, string_t &buf)
+{
+  auto it = Rust::DECOMPOSITION_MAP.find (c);
+  if (it != Rust::DECOMPOSITION_MAP.end ())
+    {
+      string_t decomped = it->second;
+      for (codepoint_t cp : decomped)
+	{
+	  recursive_decomp_cano (cp, buf);
+	}
+    }
+  else
+    buf.push_back (c);
+}
+
+string_t
+decomp_cano (string_t s)
+{
+  // TODO: Algorithmic lookup for Hangul
+  string_t buf;
+  for (codepoint_t c : s)
+    recursive_decomp_cano (c, buf);
+  return buf;
+}
+
+void
+sort_cano (string_t &s)
+{
+  int cc_here, cc_prev;
+  if (s.size () == 1)
+    return;
+  for (unsigned int i = 1; i < s.size (); i++)
+    {
+      cc_here = lookup_cc (s[i]);
+      cc_prev = lookup_cc (s[i - 1]);
+      if (cc_here >= 0 && cc_prev > cc_here)
+	{
+	  // swap
+	  int tmp = s[i];
+	  s[i] = s[i - 1];
+	  s[i - 1] = tmp;
+	  if (i > 1)
+	    i -= 2;
+	}
+    }
+}
+
+string_t
+recomp (string_t s)
+{
+  // TODO: Algorithmic lookup for Hangul
+  string_t buf;
+  if (s.size () > 0)
+    {
+      int last_class = -1;
+      // int starter_pos = 0; // Assume the first character is Starter. Correct?
+      // int target_pos = 1;
```

Comment:
dead code?

Reply:
fixed in https://github.com/Rust-GCC/gccrs/pull/2463